### ISSUE TITLE
Add Support for SPDX License Expressions

### ISSUE THREAD

**Issue by abhisek** [safedep/vet#286]
`vet` should support handling license expressions such as `(MIT OR CC0-1.0)` as per SPDX license expressions rule.

License expressions
https://spdx.github.io/spdx-spec/v2.3/SPDX-license-expressions/

Example:
https://www.npmjs.com/package/type-fest

This information should be used to evaluate if a policy violates the license expression for a given OSS package

**Comment by harisudarsan1** [comment]
### Problem Statement

To enable SPDX expression validation in Safedep, we need to parse and evaluate license expressions against policy-defined rules. The solution must accurately handle boolean logic, versioning, and containment checks without generating false positives or negatives.

### Solution Approaches

### Approach 1: Leveraging Existing Go Libraries (spdx-exp)

- **Library**: [[spdx-exp (Go)](https://github.com/licenseware/spdx-exp)](https://github.com/licenseware/spdx-exp)
- **Capabilities**:
    - Validates license expressions.
    - Supports versioning for accurate SPDX license matching.
    - `Satisfy` function checks if a list of licenses satisfies a license expression.
- **Limitations**:
    - **Expression Comparison**: The `Satisfy` function compares an expression against a list of licenses, not another expression. This introduces issues when boolean logic is involved (e.g., `MIT AND Apache-2.0` is simplified to `[MIT, Apache-2.0]`, losing the logical structure).
    - **False Positives**: License extraction from expressions ignores boolean operators, leading to incorrect validation.
- **Key Issue**: While version-aware, the inability to compare two full expressions directly (with boolean logic intact) makes this unsuitable for robust policy validation.

### Approach 2: Custom Parsing and Comparison Utilities

- **Reference Library**: [[license-expression (Python)](https://github.com/nexB/license-expression)](https://github.com/nexB/license-expression)
- **Advantages**:
    - Better API for parsing and comparing expressions.
    - Handles boolean logic and expression normalization.
- **Limitations**:
    - Uses substring matching during containment checks, leading to false negatives (e.g., `MIT AND Apache-2.0` != `Apache-2.0 AND MIT`).
    - Version mismatches and operand reordering can cause validation failures.
- **Key Issue**: Simplification and substring-based containment checks do not fully respect SPDX versioning and boolean logic integrity.

### Recommended Approach: Custom Utilities for SPDX Expression Comparison

- **Why Custom Utilities?**
    - Existing solutions fall short in handling boolean logic, versioning, and operand reordering.
    - A purpose-built parser and comparator can ensure accurate containment checks.
- **Design Goals**:
    1. **Boolean Logic Preservation**: Ensure operators (`AND`, `OR`, `WITH`) are respected during containment checks.
    2. **Version Awareness**: Handle SPDX versioning explicitly to avoid false positives/negatives.
    3. **Symmetry**: Ensure `MIT AND Apache-2.0` == `Apache-2.0 AND MIT`.
    4. **Scalability**: Optimize for large-scale license expressions and policies.
- **Implementation Plan**:
    1. **Parser**: Build a parser to convert SPDX expressions into an abstract syntax tree (AST) representing the full logical structure.
    2. **Comparator**: Develop a comparator that recursively evaluates AST nodes, ensuring containment checks respect boolean relationships and versions.
    3. **Validation API**: Expose a high-level API for expression-to-expression comparison, policy enforcement, and version validation.
- **Fallback Strategy**:
    - Initially leverage `spdx-exp` for basic version-aware license extraction and validation.
    - Gradually replace core comparison logic with the custom utility to minimize false positives.

### Conclusion

The most reliable solution involves developing custom SPDX parsing and comparison utilities that accurately handle boolean logic, versioning, and operand reordering. This approach ensures that Safedep's policy enforcement mechanism is robust, scalable, and precise, avoiding the pitfalls of existing libraries.

**Comment by abhisek** [comment]
@harisudarsan1 Thanks for the analysis.

> The solution must accurately handle boolean logic, versioning, and containment checks without generating false positives or negatives.

We can achieve this iteratively as well. May be start with base cases and then handle additional cases in future iterations (PR). What do you suggest are the base cases for this feature to be useful?

> `Satisfy` function checks if a list of licenses satisfies a license expression.

I think this is good for us. Consider the case of https://www.npmjs.com/package/type-fest, its license expression is `(MIT OR CC0-1.0)`. As a `vet` user, I should be able to match it using the following policy (`--filter` in `vet`).

```
vet scan --purl pkg:/npm/type-fest@4.31.0 --filter 'licenses.exists(p, p == "MIT")'
```

Currently, `vet` does not match this. It needs the full license string to match. Example:

```
vet scan --purl pkg:/npm/type-fest@4.31.0 --filter 'licenses.exists(p, p == "CC0-1.0 OR MIT")'
```

We need to reach a stage where `vet` can match licenses that contains license expressions instead of a single license such as what we see for `type-fest`.

Given this, I don't think the following limitation of `spdx-exp` Go library is a problem. Since `vet` supports a policy language, we can expect users to use the policy language instead of using an SPDX license expression as input. We need to handle license expressions for OSS packages such as `type-fest`.

What do you think?

**Comment by harisudarsan1** [comment]
Yeah even I thought the same. We can slowly build support for spdx expression based filtering. I shall proceed with this implementation.
